lib/resolve-external.js

```diff
@@ -75,7 +75,11 @@ function crawl (obj, path, $refs, options, external, seen) {
       if (external && $Ref.is$Ref(obj)) {
         /* Correct the reference in the external document so we can resolve it */
         let withoutHash = url.stripHash(path);
-        obj.$ref = url.toFileSystemPath(withoutHash) + obj.$ref;
+        let externalRef = withoutHash + obj.$ref;
+        if (!url.isHttp(withoutHash)) {
+          externalRef = url.toFileSystemPath(withoutHash) + obj.$ref;
+        }
+        obj.$ref = externalRef;
```

Comment:
Larger question: Is there any danger of this being applied to protocols that _aren't_ HTTP, but are networked?  (Caveat: I didn't read `url.isHttp()` so that might be more sophisticated than I assume.)

Tiny nit (take it or leave it): You might consider making `withoutHash` and `externalRef` `const`.

```TypeScript
        const withoutHash = url.stripHash(path);
        const externalRef = url.isHttp(withoutHash) ? withoutHash : url.toFileSystemPath(withoutHash);
        obj.$ref = externalRef + obj.$ref;
```

Reply:
`url.isHttp()` basically regex matches the string using `protocolPattern = /^(\w{2,}):\/\//i,` to get the protocol then 
```
exports.isHttp = function isHttp (path) {
  let protocol = url.getProtocol(path);
  if (protocol === "http" || protocol === "https") {
    return true;
  }
  else if (protocol === undefined) {
    // There is no protocol.  If we're running in a browser, then assume it's HTTP.
    return process.browser;
  }
  else {
    // It's some other protocol, such as "ftp://", "mongodb://", etc.
    return false;
  }
};
```

so that means yes, there is that danger. `toFileSystemPath` removes `file://` if it exists, decodes some url patterns and then if its a windows os, it replaces forward slashes with back slashes. I don't know that that would really mess up something that is networked, but I could be missing something. Maybe the decoding part might be a problem?


```
exports.toFileSystemPath = function toFileSystemPath (path, keepFileProtocol) {
  // Step 1: `decodeURI` will decode characters such as Cyrillic characters, spaces, etc.
  path = decodeURI(path);

  // Step 2: Manually decode characters that are not decoded by `decodeURI`.
  // This includes characters such as "#" and "?", which have special meaning in URLs,
  // but are just normal characters in a filesystem path.
  for (let i = 0; i < urlDecodePatterns.length; i += 2) {
    path = path.replace(urlDecodePatterns[i], urlDecodePatterns[i + 1]);
  }

  // Step 3: If it's a "file://" URL, then format it consistently
  // or convert it to a local filesystem path
  let isFileUrl = path.substr(0, 7).toLowerCase() === "file://";
  if (isFileUrl) {
    // Strip-off the protocol, and the initial "/", if there is one
    path = path[7] === "/" ? path.substr(8) : path.substr(7);

    // insert a colon (":") after the drive letter on Windows
    if (isWindows && path[1] === "/") {
      path = path[0] + ":" + path.substr(1);
    }

    if (keepFileProtocol) {
      // Return the consistently-formatted "file://" URL
      path = "file:///" + path;
    }
    else {
      // Convert the "file://" URL to a local filesystem path.
      // On Windows, it will start with something like "C:/".
      // On Posix, it will start with "/"
      isFileUrl = false;
      path = isWindows ? path : "/" + path;
    }
  }

  // Step 4: Normalize Windows paths (unless it's a "file://" URL)
  if (isWindows && !isFileUrl) {
    // Replace forward slashes with backslashes
    path = path.replace(forwardSlashPattern, "\\");

    // Capitalize the drive letter
    if (path.substr(1, 2) === ":\\") {
      path = path[0].toUpperCase() + path.substr(1);
    }
  }

  return path;
};
```

Reply:
I did take the nit!

Reply:
I was mostly thinking of stuff like the `stoplight://` protocol we invented.  (I doubt anybody would be using `ftp://` or anything like that.)  Is it the case that anything with a protocol that _isn't_ `file://` is something we expect to involve network access?

Reply:
I attempted to add a test for `stoplight://` but I am getting errors just reading in the file and this new code is never hit: `SyntaxError: Unable to resolve $ref pointer "stoplight://resources/overrides/aBcdEfG"
    at onError (lib/parse.js:85:20)` so I don't believe its supported right now. Maybe @P0lip might know why we don't?

Reply:
we don't have a registered resolver in the package itself, we only have it in our internal repo, so those URIs are not recognized